src/coreclr/jit/runtimelookup.cpp

```diff
@@ -0,0 +1,434 @@
+//    runtime, and not at compile - time. See ASCII block diagrams in comments below for
+//    better understanding how this phase expands runtime lookups.
+//
+PhaseStatus Compiler::fgExpandRuntimeLookups()
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHaveExpRuntimeLookup())
+    {
+        // The method being compiled doesn't have expandable runtime lookups. If it does
+        // and doesMethodHaveExpRuntimeLookup() still returns false we'll assert in LowerCall
+        return result;
+    }
+
+    INDEBUG(bool irIsPrinted = false);
+
+    // Find all calls with GTF_CALL_M_EXP_RUNTIME_LOOKUP flag
+    for (BasicBlock* block : Blocks())
+    {
+        for (Statement* const stmt : block->Statements())
+        {
+            if ((stmt->GetRootNode()->gtFlags & GTF_CALL) == 0)
+            {
+                // TP: Stmt has no calls - bail out
+                continue;
+            }
+
+            for (GenTree* const tree : stmt->TreeList())
+            {
+                // We only need calls with IsExpRuntimeLookup() flag
+                if (!tree->IsCall() || !tree->AsCall()->IsExpRuntimeLookup())
+                {
+                    continue;
+                }
+                assert(tree->IsHelperCall());
+                JITDUMP("Expanding runtime lookup for [%06d] in " FMT_BB ":\n", dspTreeID(tree), block->bbNum)
+                DISPTREE(tree)
+                JITDUMP("\n")
+
+                GenTreeCall* call = tree->AsCall();
+
+                // Clear ExpRuntimeLookup flag so we won't miss any runtime lookup that needs partial expansion
+                call->ClearExpRuntimeLookup();
+
+                if (call->IsTailCall())
+                {
+                    // It is very unlikely to happen and is impossible to represent in C#
+                    continue;
+                }
+
+                assert(call->gtArgs.CountArgs() == 2);
+                // The call has the following signature:
+                //
+                //   type = call(genericCtx, signatureCns);
+                //
+                void* signature = GetConstantPointer(this, call->gtArgs.GetArgByIndex(1)->GetNode());
+                if (signature == nullptr)
+                {
+                    // Technically, it is possible (e.g. it was CSE'd and then VN was erased), but for Debug mode we
+                    // want to catch such cases as we really don't want to emit just a fallback call - it's too slow
+                    assert(!"can't restore signature argument value");
+                    continue;
+                }
+
+#ifdef DEBUG
+                // Print full IR before any changes we're goint to make
+                if (!irIsPrinted && verbose)
+                {
+                    irIsPrinted = true;
+                    printf("\n*************** Before fgExpandRuntimeLookups()\n");
+                    fgDispBasicBlocks(true);
+                    printf("\n");
+                }
+#endif
```

Comment:
Shouldn't this be handled as part of the normal phase handling logic?

Reply:
Probably? removed